compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollAndDistributeWorkgroup.cpp

```diff
@@ -0,0 +1,503 @@
+///   loop
+class AMDAIEUnrollWorkgroupLoops : public OpRewritePattern<scf::ForOp> {
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+
+  /// Collect the internal logical objectfifos within a for op.
+  void collectInternalLogicalObjectFifos(
+      scf::ForOp forOp, DenseSet<AMDAIE::LogicalObjectFifoFromMemrefOp>
+                            &internalLogicalObjectFifos) const {
+    forOp.walk([&](AMDAIE::DmaCpyNdOp dmaOp) {
+      auto sourceObjectFifo = dmaOp.getSourceObjectFifo();
+      auto sourceMemspace = sourceObjectFifo.getMemrefType().getMemorySpace();
+      auto targetObjectFifo = dmaOp.getTargetObjectFifo();
+      auto targetMemspace = targetObjectFifo.getMemrefType().getMemorySpace();
+
+      // These are the checks to determine internal logical objectfifos.
+      // Internal means more local in the memory hierarchy, for example L1 is
+      // more local than L2 and if we have a dma copy from L2 to L1, the L1
+      // logical objectfifo is an internal one. This can also be thought about
+      // as the non-boundary logical objectfifos.
+      //   1) If source is on global memory, target is an interal logical
+      //   objectfifo.
+      //   2) If target is on global memory, source is an interal logical
+      //   objectfifo.
+      //   3) If target memory space is more local than source memory space,
+      //   target is an interal logical objectfifo.
+      //   4) If target memory space is less local than source memory space,
+      //   source is an interal logical objectfifo.
+      if (!sourceMemspace) {
+        internalLogicalObjectFifos.insert(targetObjectFifo);
+      } else if (!targetMemspace) {
+        internalLogicalObjectFifos.insert(sourceObjectFifo);
+      } else if (dyn_cast<IntegerAttr>(targetMemspace).getInt() >
+                 dyn_cast<IntegerAttr>(sourceMemspace).getInt()) {
+        // More locally, towards AIE
+        internalLogicalObjectFifos.insert(targetObjectFifo);
+      } else if (dyn_cast<IntegerAttr>(targetMemspace).getInt() <
+                 dyn_cast<IntegerAttr>(sourceMemspace).getInt()) {
+        // More globally, away from AIE
+        internalLogicalObjectFifos.insert(sourceObjectFifo);
+      }
+      return WalkResult::advance();
+    });
+  }
+
+  /// Hoist dma ops outside the scf.for operation if none of the dma op's
+  /// operands depend on the scf.for's induction variable or any operation
+  /// within the scf.for's scope.
+  LogicalResult hoistDmaOps(PatternRewriter &rewriter, scf::ForOp forOp) const {
+    bool hoistHappened{false};
+
+    // Create utility function to check whether an operand depends on a scf.for
+    // induction variable or a value within the scf.for's scope
+    auto dependsOnLoop = [&](OpOperand &operand) -> bool {
+      Operation *op = operand.get().getDefiningOp();
+      if (!op) return operand.get() == forOp.getInductionVar();
+
+      // Check fo induction var and whether the parent scf.for is the same as
```

Comment:
`Check fo induction` -> `Check for induction` ?

Reply:
Thanks, done.